File: homeworks/Nigmatullin/stack/Stack.cpp
#include <iostream>

using namespace std;

const int N = 100;

struct stack
{
	int data[N];
	int current = -1;
};

void push(stack &s, int el)
{
	if (s.current < N - 1)
	{
		s.data[++s.current] = el;
	}
}

bool isEmpty(stack &s)
{
	return !(s.current > -1);
}

int pop(stack &s)
{
	if (!isEmpty(s))
	{
		return s.data[s.current--];
	}
	else
	{
		return -1;
	}
}

void empty(stack &s)
{
	s.current = -1;
}



int main()
{

	stack s;
	push(s, 10);

	for (int i = 0; i < 10; i++)
	{
		push(s, i);
	}
	bool ie = isEmpty(s);
	int a = pop(s);

	empty(s);

	ie = isEmpty(s);


	return 0;
}

File: homeworks/Galeev/Second_semester/InP/Stack.pas
unit Stack;

  INTERFACE

  const
    CN = 100;
  
  type
    T = integer;
    index = 0..CN;
  
    tStack = record
      data: array[1..CN] of T;
      current: index;
    end;

  procedure push(var s: tStack; el: T);
  procedure empty(var s: tStack);
  function pop(s: tStack): T;

  IMPLEMENTATION  

  procedure push(var s: tStack; el: T);
  begin
    if (s.current < CN) then
    begin
      inc(s.current);
      s.data[s.current] := el;
    end;
  end;

  function isEmpty(s: tStack): boolean;
  begin
    result := not(s.current > 0);
  end;

  function pop(s: tStack): T;
  begin
    if not isEmpty(s) then
    begin
      result := s.data[s.current];
      dec(s.current);
    end
    else
      result := -1;
  end;

  procedure empty(var s: tStack);
  begin
    s.current := 0;
  end;
end.